File: ios/Sources/KlinikPhotosFeature/FlaggedPhotosModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum FlaggedPhase: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case notPermitted
    case failed(String)
}

public struct FlaggedState: Sendable, Equatable {
    public var phase: FlaggedPhase = .loading
    public var photos: [FlaggedPhoto] = []
    public var busyId: String?
    public var error: String?

    public init() {}
}

/**
 * Photographs the pre-read marked for a clinician (spec M5, M7).
 *
 * A worklist, oldest first: a queue sorted newest-first is a queue whose oldest
 * item waits forever, and the oldest item here is somebody who photographed a
 * wound three days ago because it worried them.
 *
 * The pre-read is a flag, never a diagnosis. Nothing on this screen says what
 * is wrong — it says a machine noticed something and a person should look,
 * which is the only claim the model is allowed to make.
 */
@MainActor
public final class FlaggedPhotosModel {
    private let api: PhotosAPI
    private var state = FlaggedState()

    public init(api: PhotosAPI) {
        self.api = api
    }

    public func currentState() -> FlaggedState { state }

    public func load() async {
        do {
            state.photos = try await api.flagged()
            state.phase = state.photos.isEmpty ? .empty : .loaded
        } catch APIError.forbidden {
            state.phase = .notPermitted
        } catch let error as APIError {
            state.phase = .failed(L10n.message(for: error))
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    /// Asks for a pre-assessment of one photograph.
    ///
    /// Sends a clinical photograph to a third party, which is why it is a press
    /// and not something the gallery does on scroll.
    public func assess(_ photoId: String) async -> AssessmentSkip? {
        state.busyId = photoId
        state.error = nil

        defer { state.busyId = nil }

        do {
            let assessment = try await api.assess(photoId)

            if let index = state.photos.firstIndex(where: { $0.id == photoId }) {
                state.photos[index] = state.photos[index].reassessed(as: assessment.photo)
            }

            return assessment.skippedReason
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }

        return nil
    }

    /// Oldest first — the point of a worklist.
    public func ordered() -> [FlaggedPhoto] {
        state.photos.sorted { $0.takenAt < $1.takenAt }
    }
}

File: android/feature/photos/src/main/kotlin/xyz/klinik/feature/photos/FlaggedPhotosModel.kt
package xyz.klinik.feature.photos

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.FlaggedPhoto
import xyz.klinik.network.PhotosApi
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface FlaggedPhase {
    data object Loading : FlaggedPhase
    data object Loaded : FlaggedPhase

    /** Nothing flagged. The good state, and worth saying rather than drawing blank. */
    data object Empty : FlaggedPhase

    /** Not a failure: the photo endpoints need `photos.read`. */
    data object NotPermitted : FlaggedPhase
    data class Failed(val message: UiText) : FlaggedPhase
}

data class FlaggedPhotosState(
    val phase: FlaggedPhase = FlaggedPhase.Loading,
    /** Oldest first: the one that has been waiting longest for a look. */
    val photos: List<FlaggedPhoto> = emptyList(),
    val busyId: String? = null,
    val error: UiText? = null,
)

/**
 * The wound photographs an assessment thought somebody should look at (M5).
 *
 * A flag, never a diagnosis. The findings come from a closed vocabulary — the
 * server drops anything outside it rather than passing it through — and the
 * disclaimer travels with every row, because a list of photographs headed
 * "redness, discharge" reads as a conclusion unless it says otherwise.
 *
 * Oldest first, deliberately. This is a queue of work, and the photograph that
 * has been waiting since Tuesday is the one somebody should see.
 */
class FlaggedPhotosModel(private val api: PhotosApi) {
    private val _state = MutableStateFlow(FlaggedPhotosState())
    val state: StateFlow<FlaggedPhotosState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = FlaggedPhase.Loading)

        try {
            val photos = api.flagged().sortedBy { it.takenAt }

            _state.value = FlaggedPhotosState(
                phase = if (photos.isEmpty()) FlaggedPhase.Empty else FlaggedPhase.Loaded,
                photos = photos,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.Forbidden) {
                    FlaggedPhase.NotPermitted
                } else {
                    FlaggedPhase.Failed(messageFor(error))
                },
            )
        }
    }

    /**
     * Runs the assessment again on one photograph.
     *
     * The row is replaced with what came back rather than removed: a
     * re-assessment that found nothing is an answer about that photograph, and
     * a clinician who watched it vanish would not know which it was.
     */
    suspend fun reassess(photo: FlaggedPhoto): Boolean {
        _state.value = _state.value.copy(busyId = photo.id, error = null)

        return try {
            val assessment = api.assess(photo.id)

            _state.value = _state.value.copy(
                // The assessment answers about the photograph; the patient it
                // belongs to is not part of that answer and is kept from the
                // row we already have.
                photos = _state.value.photos.map {
                    if (it.id == photo.id) {
                        it.copy(
                            aiReviewSuggested = assessment.photo.aiReviewSuggested,
                            aiFindings = assessment.photo.aiFindings,
                            aiAssessedAt = assessment.photo.aiAssessedAt,
                        )
                    } else {
                        it
                    }
                },
                busyId = null,
            )

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busyId = null, error = messageFor(error))
            false
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
